File: sugar.go
package main

import (
	_ "embed"
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/joho/godotenv"
	openai "github.com/sashabaranov/go-openai"
)

//go:embed .env
var embeddedEnvContent string

func loadDotenv() {
	// Try to load dotenv from file in directory
	err := godotenv.Load()

	// If dotenv file is not found in directory, use embedded content
	if err != nil {
		lines := strings.Split(embeddedEnvContent, "\n")
		for _, line := range lines {
			parts := strings.SplitN(line, "=", 2)
			if len(parts) == 2 {
				key := parts[0]
				value := parts[1]
				os.Setenv(key, value)
			}
		}
	}
}

func makeSentenceMoreFriendly(sentence string, apiKey string) (string, error) {
	// Create client with API key
	client := openai.NewClient(apiKey)

	// Call the function that returns the chat completion
	response, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: openai.GPT3Dot5Turbo,
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleSystem,
					Content: "Make the user-provided sentence less blunt, and a bit more friendly, but without exaggeration",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: sentence,
				},
			},
		},
	)

	if err != nil {
		fmt.Printf("ChatCompletion error: %v\n", err)
		return "", err
	}

	return strings.TrimSpace(response.Choices[0].Message.Content), nil
}

func main() {

	// Parse sentence to sweeten
	args := os.Args[1:]
	if len(args) == 0 {
		fmt.Println("Usage: ./sugar <blunt_sentence>")
		return
	}
	bluntSentence := strings.Join(args, " ")

	// Load OpenAI API key
	loadDotenv()
	apiKey := os.Getenv("OPENAI_API_KEY")

	// Sweeten the sentence with GPT
	friendlySentence, err := makeSentenceMoreFriendly(bluntSentence, apiKey)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// The only program output is the rewritten sentence
	fmt.Println(friendlySentence)
}

File: sugar.py
#!/usr/bin/env python

import os
import sys
import argparse
from dotenv import load_dotenv
from openai import OpenAI

# Load environment variables from .env file
# https://github.com/pyinstaller/pyinstaller/issues/5522#issuecomment-770858489
extDataDir = os.getcwd()
if getattr(sys, 'frozen', False):
    extDataDir = sys._MEIPASS
load_dotenv(dotenv_path=os.path.join(extDataDir, '.env'))

# Get the OpenAI API key from environment variables
client = OpenAI(
    # This is the default and can be omitted
    api_key=os.environ.get("OPENAI_API_KEY"),
)

def make_sentence_more_friendly(sentence):
    prompt = "Make the user-provided sentence less blunt, and a bit more friendly, but without exaggeration"
    response = client.chat.completions.create(
        messages=[
            {
                "role": "system",
                "content": prompt
            },
            {
                "role": "user",
                "content": sentence,
            }
        ],
        model="gpt-3.5-turbo",
    )
    return response.choices[0].message.content.strip()

def main():
    parser = argparse.ArgumentParser(description="Make a blunt sentence a bit more friendly")
    parser.add_argument("sentence", help="The blunt sentence to be made more friendly")
    args = parser.parse_args()

    friendly_sentence = make_sentence_more_friendly(args.sentence)
    print(friendly_sentence)

if __name__ == "__main__":
    main()
